src/Psalm/Internal/Analyzer/Statements/Expression/Call/FunctionCallReturnTypeFetcher.php

```diff
@@ -58,7 +59,26 @@ public static function fetch(
         $stmt_type = null;
         $config = $codebase->config;
 
-        if ($codebase->functions->return_type_provider->has($function_id)) {
+        if ($stmt->isFirstClassCallable()) {
```

Comment:
Maybe this should be after the return type provider check, would that work? The risk otherwise, you prevent using return type providers when using first class callables (this limits what plugins can do, and even some Psalm internal providers)

Reply:
If I understand correctly, we don't want to use the return type provider as we're not looking for the return type identified by `$function_id`, rather returning a closure instead.

Reply:
Oh yeah, of course. We'll have to check that the type providers are correctly called when resolving the closure call then